src/script/sigcache.h

```diff
@@ -16,13 +17,40 @@ static const unsigned int DEFAULT_MAX_SIG_CACHE_SIZE = 40;
 
 class CPubKey;
 
+/**
+ * A thread safe map which caches mid state witness signature hash calculation by transaction id
+ */
+class CachedHashesMap
+{
+private:
+    std::map<uint256, CachedHashes> map;
+    CCriticalSection cs;
+public:
+    bool TryGet(uint256 txId, CachedHashes* hashes)
+    {
+        LOCK(cs);
+        if(!map.count(txId))
+            return false;
+        *hashes = map[txId];
+        return true;
```

Comment:
> What do you mean by "fixed early-exit" ?

I accidentally inverted the early exit logic,  I since fixed it,  but left the edit in case you read it in an email and followed up.